zypp-core/zyppng/base/private/linuxhelpers_p.h

```diff
@@ -17,15 +17,40 @@ namespace zyppng {
     return std::string( strBuf.data() );
   }
 
-  template<typename Fun, typename... Args >
-  auto eintrSafeCall ( Fun &&function, Args&&... args ) {
+
+  /*!
+   * Extended version of \ref eintrSafeCall
+   *
+   * Calls a given \a function in a \a EINTR safe way. The \a restartCb is
+   * invoked after every time we have to restart \a function because of EINTR, it can be used to
+   * reset additional errors:
+   * \code
+   * FILE *myFile;
+   * // open file
+   * // ...
+   * eintrSafeCall( ::getline, [&](){ ::clearerr(myFile); }, myFile );
+   * \endcode
+   */
+  template<typename Fun, typename RestartCb, typename... Args >
+  auto eintrSafeCallEx ( Fun &&function, RestartCb &&restartCb, Args&&... args ) {
     int res;
-    do {
-      res = std::forward<Fun>(function)( std::forward<Args>(args)... );
-    } while ( res == -1 && errno == EINTR );
+    while ( true ) {
+      errno = 0;
+      res = function( args... );
```

Comment:
```suggestion
      res = std::forward<Fun>(function)( std::forward<Args>(args)... );
```

Reply:
What if we get a rvalue reference ? The calling function would be free to move the object away , so in a 2nd iteration we would have undefined behavior?